### PR TITLE
#23: Add logout button

### PR DESCRIPTION
**Pull request by Maxime-Oriol**
This feature add a logout button. For the moment it does nothing except push SignInViewController

### CODE REVIEW

File: TreeTracker/UI/Views/Home/HomeViewModel.swift [hunk 1]
```diff
@@ -58,6 +59,11 @@ class HomeViewModel {
     func addTreeSelected() {
         coordinatorDelegate?.homeViewModel(self, didSelectAddTreeForPlanter: planter)
     }
+
+    func logoutPlanter() {
+// TODO: Sync & Clear session before logout
```
Comment (AlexCornforth) [inline TreeTracker/UI/Views/Home/HomeViewModel.swift]: Like your thinking but we shouldn't need to Sync & Clear here. All the users data will already be saved in core data and should persist there between sessions.

File: TreeTracker/UI/Views/Home/HomeViewController.swift [hunk 2]
```diff
@@ -97,6 +103,10 @@ private extension HomeViewController {
     @IBAction func uploadsButtonPressed() {
         viewModel?.uploadListSelected()
     }
+
+    @objc func logoutButtonPressed() {
+        self.viewModel?.logoutPlanter()
```
Comment (AlexCornforth) [inline TreeTracker/UI/Views/Home/HomeViewController.swift]: We don't need the `self` keyword here. I know it's valid code but I think we should be consistent throughout the codebase with this. Swift encourages us to omit `self` and only use it when the compiler requires it.

File: TreeTracker/UI/Views/Home/HomeViewController.swift [hunk 3]
```diff
@@ -84,6 +84,12 @@ class HomeViewController: UIViewController, AlertPresenting {
     override func viewDidLoad() {
         super.viewDidLoad()
         viewModel?.fetchTrees()
+        self.showLogoutButton()
+    }
+
+    private func showLogoutButton() {
+        let logoutButton = UIBarButtonItem(title: "Log Out", style: .plain, target: self, action: #selector(logoutButtonPressed))
```
Comment (AlexCornforth) [inline TreeTracker/UI/Views/Home/HomeViewController.swift]: Would be good to add the "Log Out" text to Localizable.strings then use the SwiftGen generated `L10n` value here.

File: TreeTracker/UI/Views/Home/HomeViewController.swift [hunk 4]
```diff
@@ -84,6 +84,12 @@ class HomeViewController: UIViewController, AlertPresenting {
     override func viewDidLoad() {
         super.viewDidLoad()
         viewModel?.fetchTrees()
+        self.showLogoutButton()
+    }
+
+    private func showLogoutButton() {
```
Comment (AlexCornforth) [inline TreeTracker/UI/Views/Home/HomeViewController.swift:93]: For consistency with the other ViewControllers can we create a private extension for this similar to the button actions? I'm sure as time goes on we will have more private functions we will want to drop in here.
 e.g.

```
// MARK: - Private 
private extension HomeViewController { 
    func showLogoutButton() { ... }
}
```

File: TreeTracker/UI/Views/Home/HomeViewController.swift [hunk 5]
```diff
@@ -84,6 +84,12 @@ class HomeViewController: UIViewController, AlertPresenting {
     override func viewDidLoad() {
         super.viewDidLoad()
         viewModel?.fetchTrees()
+        self.showLogoutButton()
```
Comment (AlexCornforth) [inline TreeTracker/UI/Views/Home/HomeViewController.swift]: Same as my other comment, can we remove the `self` keyword here.

File: TreeTracker/UI/Navigation/Home/HomeCoordinator.swift [hunk 6]
```diff
@@ -111,6 +127,10 @@ extension HomeCoordinator: HomeViewModelCoordinatorDelegate {
     func homeViewModel(_ homeViewModel: HomeViewModel, didSelectUploadListForPlanter planter: Planter) {
         showUploadList(planter: planter)
     }
+
+    func homeViewModel(_ homeViewModel: HomeViewModel, didLogoutPlanter planter: Planter) {
+        logoutPlanter(planter: planter)
```
Comment (AlexCornforth) [inline TreeTracker/UI/Navigation/Home/HomeCoordinator.swift]: The HomeCoordinatorDelegate has a function called `homeCoordinatorDidLogout(_ homeCoordinator: HomeCoordinator)`, you should just need to call that here.

The RootCoordinator responds to this and restarts the AuthenticationCoordinator when it is called.

**Review by AlexCornforth — CHANGES_REQUESTED** [review]
Great work on this Maxime! I've left comments on things to resolve. 

In regards to using `self` vs not using `self` I would be interested to hear your thoughts if you prefer one way over the other. Personally I prefer to only use it when the compiler requires.

**Comment by Maxime-Oriol** [comment]
I used to use `self` to identify properties easily but due to your comment I looked up for best practice. It appears Swift documentation recommend to use self only when it's necessary.
Maybe I'll use `self` few times because of bad habits.

**Review by AlexCornforth — APPROVED** [review]
Nice work 👍

**Pull request merged**